main/lsp/lsp_helpers.cc

```diff
@@ -332,7 +332,7 @@ optional<string> findDocumentation(string_view sourceCode, int beginIndex) {
                 // Start of sig block
                 && !(absl::StartsWith(line, "sig do") || absl::StartsWith(line, "sig(:final) do"))
                 // Invalid end keyword
-                && !absl::StartsWith(line, "end")) {
+                && !(absl::StartsWith(line, "end") && absl::EndsWith(line, "end"))) {
```

Comment:
Since `line` has been initialized with `StripAsciiWhitespace`, we can just check `line != "end"`.

Doing it the way this PR currently implements it with "StartsWith + EndsWith" means that the bug in #6676 could still happen for a signature like

```ruby
sig do
  params(
    end_time: Time_end
  ).void
end
def foo(end_time)
end
```

Reply:
Good point. My mistake, thank you for reviewing this.

Reply:
no worries, this code is cursed anyways. Thanks for the contribution!

Reply:
Just out of curiosity, is there any particular reason why you need to handle this "invalid end keyword" case here? 

I can think of somebody breaking current logic just by putting something like an `end # random` line somewhere in the `sig` block.

Reply:
The fundamental problem here is that Sorbet doesn't parse comments in its AST.

https://github.com/sorbet/sorbet/blob/ff9bec08e7c2f5404f326534d52b94002e39765e/parser/parser/cc/lexer.rl#L275-L281

There is a callback that could be implemented if Sorbet wanted to start doing that, but due to memory constraints running Sorbet over Stripe's main Ruby codebase, we've been reluctant to do anything which would cause Sorbet to use more memory.

If Sorbet were able to access the list of already-parsed comments in its main pipeline, we would not need to resort to ad hoc string scanning inside LSP helpers like this.

**tl;dr**: The current solution is a compromise so that Sorbet can use substantially less memory while sometimes failing to find a documentation comment and sometimes finding the wrong comment. Given that this information is not load boarding, a best-effort solution is mostly fine for now.

Reply:
Thank you for such a detailed reply! 

I pushed the requested changes.